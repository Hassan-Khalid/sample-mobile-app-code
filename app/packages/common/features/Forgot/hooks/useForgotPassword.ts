import { useCallback, useEffect, useState } from 'react';
import { ThemedButtonProps } from 'common/declarations/types';
// @ts-ignore
import { forgotPassword } from '../../../../apollo/services';
import { useToast } from "react-native-toast-notifications";
import { NavigationService } from '../../../services';
import SCREENS from '../../../constants/Screens';
import Util from '../../../util/Util';

export enum _state {
  EMAIL = 'EMAIL',
  RESEND = 'RESEND'
}
const useForgotPassword = ()=> {
  const [email, setEmail] = useState<string>('test@test.com')
  const [emailError, setEmailError] = useState(false);
  const [loading,setLoading ] = useState(false);
  const [state, setComponentState] = useState(_state.EMAIL);
  const componentState = {
    'EMAIL': {
      heading: 'login.forgotPassword.title',
      subHeading: 'login.forgotPassword.message',
      button: 'login.forgotPassword.reset'
    },
    'RESEND': {
      heading: 'login.forgotPassword.resetSuccess.message',
      subHeading: `An email has been sent to ${email}, check the inbox and click the reset link.`,
      button: 'login.thanks',
    }
  }
  const [screenState, setScreenState] = useState(componentState.EMAIL)
  const { mutation } = forgotPassword();
  const toast = useToast();
  const onEmailInputChange = (email: string) => {
    setEmail(email)
  }
  const validateEmail = () => {
    if (Util.validateEmail(email) || !email) {
      setEmailError(false);
    } else {
      setEmailError(true);
    }
  }
  
  const onBlur = () =>{
    validateEmail();
  }
  
  const onPressHandler = () => {
    setLoading(true)
    if(!emailError) {
      mutation({ email: String(email).toLowerCase() }
      ).then((response) => {
        if (response.data.forgotPassword) {
          setComponentState(_state.RESEND)
          setLoading(false)
          toast.show(response.data.forgotPassword, {
            duration: 2000,
            type: "success",
            placement: "top",
            animationType: "slide-in",
          })
        }
      }).catch((e) => {
        setLoading(false)
        toast.show(e.graphQLErrors[0].message, {
          duration: 2000,
          type: "danger",
          placement: "top",
          animationType: "slide-in",
        })
        // eslint-disable-next-line no-console
        console.log('res', e.graphQLErrors[0].message)
      })
    }
  }
  
  const checkButtonState = () => {
    if ((email && !emailError) || state === _state.RESEND) {
      return false
    }
    return true
  }
  const setstate = () => {
    if (state === _state.EMAIL) {
      setScreenState(componentState.EMAIL);
      return componentState.EMAIL.button
    } else {
      setScreenState(componentState.RESEND);
      return componentState.RESEND.button
    }
  }
  
  const onHandlerPress = () => {
    if (state === _state.EMAIL) {
      onPressHandler();
    } else {
      NavigationService.navigate(SCREENS.LOGIN);
    }
  }
  const checkLoadingState = useCallback(() => {
    return loading
  },[loading])
  const createBottomButton = ()=>{
    const bottomBtn = <ThemedButtonProps>{}
    bottomBtn.title = screenState.button;
    bottomBtn.onPress = onHandlerPress
    bottomBtn.size='large';
    bottomBtn.loading = checkLoadingState()
    bottomBtn.disabled = checkButtonState()
    
    return [bottomBtn]
}
  useEffect(() => {
    validateEmail()
  }, [email]);
  
  
  useEffect(() => {
    setstate()
  }, [state]);
  
  
  return {
    onEmailInputChange,
    emailError,
    email,
    onBlur,
    state,
    onPressHandler,
    createBottomButton,
    setComponentState,
    screenState,
    loading,
    validateEmail,
    setLoading,
    setEmailError
    
  }
}
export default useForgotPassword;