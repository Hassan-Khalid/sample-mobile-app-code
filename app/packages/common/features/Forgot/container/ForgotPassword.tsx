import React from 'react';
import { View, Image, TouchableOpacity, Text as RNText } from 'react-native';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view'
import Text from "../../../components/Text/Text";
import { COMMON_LOGOS } from 'common/assets/Images';
import styles from '../styles/ForgotPassword.styles';
import InputText from '../../../../common/components/InputText/InputText';
import useForgotPassword ,{_state} from '../hooks/useForgotPassword';
import BottomActionButton from '../../../../common/components/BottomActionButton/BottomActionButton';
import { t } from 'i18next';

function ForgotPassword() {
  const {onEmailInputChange,emailError,screenState,onPressHandler,state,email,onBlur, createBottomButton} = useForgotPassword();
  
  const renderInputView = () => {
    if(state === _state.EMAIL){
      return (
        <View>
          <View style={styles.innerHeadingContainer}>
            <Text style={styles.heading}>{screenState.heading}</Text>
            <Text style={styles.subHeading}>{screenState.subHeading}</Text>
          </View>
          <InputText
            label={'login.email.label'}
            onBlur={onBlur}
            errorMessage={'login.email.invalidEmail'}
            error={emailError}
            value={email}
            keyboardType={'email-address'}
            onChangeText={onEmailInputChange}
            placeholder={"login.email.placeholder"}
          />
        </View>
      )
    }
    else {
      return (
        <View>
          <View style={styles.innerHeadingContainer}>
            <Text style={styles.heading}>{screenState.heading}</Text>
            <RNText style={styles.subHeading}>
              <Text style={styles.subHeading}>{'login.forgotPassword.resetSuccess.resend.resendEmailSubHeading'}</Text>
              <Text style={{...styles.subHeading,...styles.boldText}}>{' ' + email + ','}</Text>
              <Text style={styles.subHeading}>{'login.forgotPassword.resetSuccess.resend.resendEmailHeading'}</Text>
            </RNText>
          <View style={styles.resendView}>
              <RNText style={styles.resendText}>{t('login.forgotPassword.resetSuccess.resend.didnotGet')}</RNText>
                <RNText>{' '}</RNText>
                 <TouchableOpacity onPress={onPressHandler} >
                <RNText style={styles.resendBtnText}>{t('login.forgotPassword.resetSuccess.resend.resendEmail')}</RNText>
              </TouchableOpacity>
          </View>
          </View>
        </View>
      )
    }
    
  }
  
  return (
    <View  style={styles.container}>
      <KeyboardAwareScrollView
        enableOnAndroid>
        <View style={styles.bannerContainer}>
          <Image source={COMMON_LOGOS.LOGIN_BG} resizeMode='cover' style={styles.topBanner}/>
        </View>
        <View style={styles.logoWrapper}>
          <Image source={COMMON_LOGOS.MAKULA_LOGO_DARK} resizeMode={'contain'}/>
        </View>
        <View style={styles.inputContainer}>
          {renderInputView()}
        </View>
       
      </KeyboardAwareScrollView>
      <BottomActionButton buttonArray={createBottomButton()}/>
    
    </View>
  );
}

export default ForgotPassword;
