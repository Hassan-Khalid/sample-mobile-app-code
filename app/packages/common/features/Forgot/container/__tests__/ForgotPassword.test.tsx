import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import ForgotPassword from '../ForgotPassword';
import useForgotPassword, { _state } from '../../hooks/useForgotPassword';

jest.mock('../../hooks/useForgotPassword');

describe('ForgotPassword component', () => {
  beforeEach(() => {
    useForgotPassword.mockReturnValue({
      onEmailInputChange: jest.fn(),
      emailError: false,
      screenState: {
        heading: 'Forgot Password',
        subHeading: 'Enter your email to reset password',
      },
      onPressHandler: jest.fn(),
      state: _state.EMAIL,
      email: 'test@example.com',
      onBlur: jest.fn(),
      createBottomButton: jest.fn(() => []),
    });
    jest.clearAllMocks()
  });
  
  it('render component', () => {
    //setup
    const wrapper = render(<ForgotPassword />);
    
    // post-conditions
    expect(wrapper.toJSON).toMatchSnapshot();
  });
  
  it('renders correctly when in email state', () => {
    const { getByText, getByPlaceholderText } = render(<ForgotPassword />);
    
    expect(getByText('Forgot Password')).toBeTruthy();
    expect(getByText('Enter your email to reset password')).toBeTruthy();
  });
  
});
