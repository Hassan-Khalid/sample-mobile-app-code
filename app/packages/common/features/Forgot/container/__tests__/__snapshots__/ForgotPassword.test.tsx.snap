// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForgotPassword component render component 1`] = `[Function]`;
